### ISSUE TITLE
Multiple problems and ideas for features

### ISSUE THREAD

**Issue by jjlankinen** [nikku/wuffle#54]
Firstly: Big thank you for providing open source version of Waffle board. 

Secondly: Sorry for a wall of text.  I do understand that this issue is not in desired format and I apologize about that. This is more of an discussion opening and general feedback for few things and it might be better to create issues separately if this discussion provides to be useful.

I finally managed to set my own prototype board up and running locally and I encountered few undocumented installation issues.

There are no mention that nodejs 10+ is required. I tried to run it with 8.12 but .finally() method is not supported in that version and therefore board only updated once (added one card only) after restarting the app. Some googling and issue was resolved but had I've known what were the requirements my life would have been a bit easier. Others might find that info valuable also as we use older nodejs for different reasons.

When using probot to create the github app the websocket secret in .env file must be updated manually as it differs from what is set in github app settings.

BASE_URL is missing in .env file if you use Probot, of course you can check the .env.example file to see how the .env should be configured.

In .env.example file there is mention that BASE_URL must match the Webhook URL which is configured in github app settings. However that setting in my case refers to smee.io (that is eventually pointed to localhost) and BASE_URL is set to http://localhost:3000. I found this a bit confusing. 

### sorting:true
Sorting, I do not understand how this should work and how it should behave when it's added to columns. I tried with sorting:true and without it and nothing changed.

I used for testing following wuffle.config.js file:

```
module.exports = {
  name: 'xxx',
  columns: [
    { name: 'Stuff and Things...', label: null, sorting: true, states: [ 'DEFAULT', 'EXTERNAL_CONTRIBUTION' ] },
    { name: 'Needs Grooming', label: 'groom', sorting: true },
    { name: 'Planned and Ready', label: 'PLANNED', sorting: true },
    { name: 'Bugs, those nasty creatures!', label: 'bug', sorting: true },
    { name: 'Do it Fred!', label: 'IN PROGRESS', sorting: true, states: [ 'IN_PROGRESS' ] },
    { name: 'Needs Review', label: 'NEEDS REVIEW', sorting: true, states: [ 'IN_REVIEW' ] },
    { name: 'WE DID IT!', label: null, closed: true, sorting: true, states: [ 'DONE' ] }
  ]
};
```
### Manually sort columns
We have quite a big backlog and icebox (because reasons) and it would be nice to be able to sort columns asc/desc by issue creation date.

### Labels
I found out that labels defined in wuffle.config.js do not appear on cards on the board which makes sense. However I would like to suggest that when using previously made labels to sort issues to different columns (like in our example above) those non state related labels should/could/would be visible (we use red bug label which is very easy to spot and now it does not appear at all), or maybe add another field that forces label to be visible ie. forceLabelVisible:true or something like that. 

I found out that it's very nice to be able to use previously made labels to sort issues as it adds awful lot of flexibility to this boards use cases.

### Rebuild app
After messing around with columns names etc. I found out (there's an another issue about this too #49) that if you change the column name you cannot see any issues in that column anymore even if the state or label are unaltered.

I would like to know how to rebuild the app and/or how to destroy the fetched data. I deleted storedump.json file from tmp folder but it did not do anything obviously. I ended up backuping .env and wuffle.config.js files and reinstalling everything from the scratch.

Is there a way to purge every fetched card or do users have to reinstall this app every time if some column name needs to be changed?

### Force Refresh cards sorting
Is there any way to force cards column based sorting ie. if I add new column called "Koalas are cute" and I want that column to be populated with issues that have label "Koala" in github they don't appear in that column after restarting the app.

Currently I have to re-add the label in github to the issue and then it will update the board properly. 

### IN_PROGRESS state
Does this state follow the same ruling in branch naming as it were in Waffle? If so it might be good idea to actually document this as I had to dig interwebs a bit to actually remember the naming convention for new branches.

**Comment by nikku** [comment]
First of all, thanks for the feedback / long letter. Find a few comments quoted, inline.

> Firstly: Big thank you for providing open source version of Waffle board.

This is not a copy of Waffle, but a thing that does exactly what Wuffle did, just better (no offense :wink:).

> There are no mention that nodejs 10+ is required. I tried to run it with 8.12 but .finally() method is not supported in that version and therefore board only updated once (added one card only) after restarting the app. Some googling and issue was resolved but had I've known what were the requirements my life would have been a bit easier. Others might find that info valuable also as we use older nodejs for different reasons.

Marked it as a requirement: https://github.com/nikku/wuffle/commit/2b828020c3f365ff48cb7c7ea78dee163088864a.

 > BASE_URL is missing in .env file if you use Probot, of course you can check the .env.example file to see how the .env should be configured.
> 
> In .env.example file there is mention that BASE_URL must match the Webhook URL which is configured in github app settings. However that setting in my case refers to smee.io (that is eventually pointed to localhost) and BASE_URL is set to http://localhost:3000. I found this a bit confusing.

`BASE_URL` is not required in development mode, as it defaults to `localhost:3000`. It becomes important for production installations only.

As mentioned in the documentation it should be absolutely possible to setup the board without jumping in development mode. From the feedback I got this is what most users do, too. If you have an idea how to improve the existing documentation for those that start in development mode, please consider filing a PR.


> Sorting, I do not understand how this should work and how it should behave when it's added to columns. I tried with sorting:true and without it and nothing changed.

How do you assume this would work?

> We have quite a big backlog and icebox (because reasons) and it would be nice to be able to sort columns asc/desc by issue creation date.

This is something that will not be supported any time soon. Issues are and will remain sorted by important -> not important. You as a user need to apply the sorting (and maintain it). However I could think of adding filters that allow you to show only particular issues, i.e. older than a month, year, light year or something.

> I found out that labels defined in wuffle.config.js do not appear on cards on the board which makes sense. However I would like to suggest that when using previously made labels to sort issues to different columns (like in our example above) those non state related labels should/could/would be visible (we use red bug label which is very easy to spot and now it does not appear at all), or maybe add another field that forces label to be visible ie. forceLabelVisible:true or something like that.
> 
> I found out that it's very nice to be able to use previously made labels to sort issues as it adds awful lot of flexibility to this boards use cases.

I don't quite see the point of this one. Why would you like to see the BUG label, if you have a dedicated column for it, called Bug?


> I deleted storedump.json file from tmp folder but it did not do anything obviously.

This is exactly how it is being done. Shut down board, delete `storedump.json`. Restart board. All sorting will be lost, for now, though.

> Currently I have to re-add the label in github to the issue and then it will update the board properly.

Delete `storedump.json` and make the board re-fetch all issues.

It will probably be easier once a solution for #49 is in place.

> Does this state follow the same ruling in branch naming as it were in Waffle? If so it might be good idea to actually document this as I had to dig interwebs a bit to actually remember the naming convention for new branches.

What is the ruling for Waffle?

**Comment by nikku** [comment]
Closing this issue for now, as I hope I was able to answer most of your questions.

Please get back to me, in case I did miss something. To do so, respond in this thread or, even better, create individual issues based on the [existing issue templates](https://github.com/nikku/wuffle/issues/new/choose).

**Comment by jjlankinen** [comment]
I know this is not Waffle boards copy, quite obvious that is (being better is matter of perspective).

If BASE_URL is not required it might be worth to mention about it in documentation or did I miss the notion. 

About sorting option, I have no idea what it should do, the description does not indicate any meaningful thing to me unfortunately, that's why I brought it up.

About the manual sorting; big icebox cannot be ordered manually within the board via drag 'n drop and I don't understand how you supposed to mark something as important in github as you say "issues are and will remain sorted by important -> not important. " I could pin 3 issues but that's just peanuts. Anyways, manual sorting is always nice to have feature which we use a lot with zenhub currently.

About labels; you asked what is the point? Well I'm UI/UX designer and while being a team lead who mostly uses kanban board to see visually github issues placed in meaningful columns it's been easier for me (as well as others) to see issues with recognizable labels (via colors) without the need to actually read the column label and the label for the column might not always be the same as the label's name. I could easily assign multiple labels to be placed to same column because reasons.

About resetting the board; Resetting board the way you said did not do anything for me and that's why I asked. I shutdown the board. Deleted the storedumb.json file and started the board and issues where still missing and they were not updated to the new column.

About IN_PROGRESS state; Let me ask you a question, what is the ruling in Wuffle and where I can find it? In Waffle it was pretty simple, if you put the issue number at the beginning of the branch name with a '-' or '_' behind it or anywhere else in the branch name with a '#' in front of it the the corresponding issue will be moved to the column specified to have IN_PROGRESS state.

Are you planning to add any features to allow proper issue manipulation within the board itself or do you want to limit it to only drag 'n drop cards between columns? It seems that re-ordering cards within the column cannot be done as well.

Anyways, thanks for answering and don't get me wrong, I'm not complaining here, it's very nice to have simple alternative kanban board that have flexibility we need (the custom columns and placing cards to columns based on label names is fooking awesome :))

**Comment by nikku** [comment]
> About sorting option, I have no idea what it should do, the description does not indicate any meaningful thing to me unfortunately, that's why I brought it up.

The board automatically sorts issues based on their declared relationships. If issue `#A` is is `Depends on #B` then it will be shown below `#B`. 

> About resetting the board; Resetting board the way you said did not do anything for me and that's why I asked. I shutdown the board. Deleted the storedumb.json file and started the board and issues where still missing and they were not updated to the new column.

This sounds like a usage / configuration issue to me. `storedump.json` is the only place that keeps state around. There is no other place the board magically fetches state from.

> About IN_PROGRESS state; Let me ask you a question, what is the ruling in Wuffle and where I can find it? In Waffle it was pretty simple, if you put the issue number at the beginning of the branch name with a '-' or '_' behind it or anywhere else in the branch name with a '#' in front of it the the corresponding issue will be moved to the column specified to have IN_PROGRESS state.

I documented our capabilities [here](https://github.com/nikku/wuffle/blob/master/docs/AUTOMATIC_CARD_MOVEMENT.md). TLDR: We support that thing.

> About labels; you asked what is the point? Well I'm UI/UX designer and while being a team lead who mostly uses kanban board to see visually github issues placed in meaningful columns it's been easier for me (as well as others) to see issues with recognizable labels (via colors) without the need to actually read the column label and the label for the column might not always be the same as the label's name. I could easily assign multiple labels to be placed to same column because reasons.

Got it. Please open an issue that requests this feature, clearly lay down your rationale and I consider adding such capability.

> Are you planning to add any features to allow proper issue manipulation within the board itself or do you want to limit it to only drag 'n drop cards between columns?

I do not plan to implement full issue manipulation on the board. Two things I consider is updating the issue title and simplifying the creation of new issues (https://github.com/nikku/wuffle/issues/65). Another thing I consider in the far future is batch updating milestone or labels for multiple selected issues.

> About the manual sorting; big icebox cannot be ordered manually within the board via drag 'n drop.

I don't understand that limitation, maybe you can elaborate on your setup a little more? At work I got a [huge backlog](https://tasks.bpmn.io/board) and do not see the need for sorting it. Instead, we employ board filtering and additional columns to plan new milestones.

**Comment by jjlankinen** [comment]
Cheers for explanation for sorting option. It did not occurred to me that it was made for declared relationships.

Thank you for adding documentation for IN_PROGRESS state behavior.

About that storedump.json issuse, I have to recheck that I did everything in correct order then. I was quite sure as well that storedump is the only place where board data is stored and I was confused because deleting did not refresh the column population after I altered column names.

There really is no need for full issue manipulation needed in the board, edit title and maybe assigning issues to developers might be enough or what ever you deem necessary. github is the main source anyways so might as well do everything there.

About that manually sorting backlog/icebox, you seem to be able to manually drag 'n drop cards within the column. Stupid me, somehow I did not managed to do it even if I was logged in. I use that to prioritize cards in few columns so top pile has higher priority etc.

Would it be possible to force re-population of columns? I have adde one column to populate it with issues that have groom label but none of the issues made before I started this board are added to the correct column. they are however added to the desired column right after I remove the groom label and re-add it back. Is this currently working as intended or is this just an oversight? I assume that when you added a possibility to set custom columns to be populated with custom labels you intended existing issues to be placed to those columns automatically also or am I grasping at straws here?
